scripts/asr_language_modelling/ngram_lm/eval_beamsearch_ngram.py

```diff
@@ -0,0 +1,186 @@
+    parser.add_argument("--beam_alpha", required=True, type=float)
+    parser.add_argument("--beam_beta", required=True, type=float)
+    parser.add_argument("--acoustic_batch_size", default=16, type=int)
+    parser.add_argument("--beam_batch_size", default=128, type=int)
+    parser.add_argument("--device", default="cuda:0", type=str)
+    parser.add_argument(
+        "--decoding_mode", choices=["greedy", "beamsearch", "beamsearch_ngram"], default="beamsearch_ngram", type=str
+    )
+    args = parser.parse_args()
+
+    logging.info(f"BEAM WIDTH : {args.beam_width}")
+    logging.info(f"BEAM ALPHA : {args.beam_alpha}")
+    logging.info(f"BEAM BETA : {args.beam_beta}")
+
+    asr_model = nemo_asr.models.EncDecCTCModelBPE.restore_from(
+        args.nemo_model_file, map_location=torch.device(args.device)
+    )
+    asr_model.preprocessor.featurizer.dither = 0
+    asr_model.preprocessor.featurizer.pad_to = 0
+    # Set model to inference mode
+    asr_model.eval()
+
+    model_tokenizer = asr_model.tokenizer
+    vocabs = list(model_tokenizer.tokenizer.get_vocab().keys())
+
+    target_transcripts = []
+    with open(args.input_manifest, 'r') as manifest_file:
+        audio_file_paths = []
+        for line in tqdm(manifest_file, desc=f"Reading Manifest {args.input_manifest} ..."):
+            data = json.loads(line)
+            target_transcripts.append(data['text'])
+            audio_file_paths.append(data['audio_filepath'])
+
+    # drop it later
+    # audio_file_paths = audio_file_paths[0:10]
+
+    if args.use_probs_cache and os.path.exists(args.probs_cache_file):
+        logging.info(f"Found a pickle file of probabilities at {args.probs_cache_file}.")
+        logging.info(f"Loading the cached pickle file of probabilities from {args.probs_cache_file}...")
+        with open(args.probs_cache_file, 'rb') as probs_file:
+            all_probs = pickle.load(probs_file)
+
+        if len(all_probs) != len(audio_file_paths):
+            raise ValueError(
+                f"The number of samples in the probabilities file '{args.probs_cache_file}' is not "
+                f"the same as the manifest file. "
+                f"You may need to delete the probabilities cached file."
+            )
+    else:
+        with torch.no_grad():
+            all_logits = asr_model.transcribe(audio_file_paths, batch_size=args.acoustic_batch_size, logprobs=True)
+        all_probs = [softmax(logits) for logits in all_logits]
+        logging.info(f"Writing pickle files of probabilities at {args.probs_cache_file}")
+        with open(args.probs_cache_file, 'wb') as f_dump:
+            pickle.dump(all_probs, f_dump)
+
+    wer_dist_greedy = 0
+    words_count = 0
+    for batch_idx, probs in enumerate(all_probs):
+        preds = np.argmax(probs, axis=1)
+        preds_tensor = torch.tensor(preds, device='cpu').unsqueeze(0)
+        pred_text = asr_model._wer.ctc_decoder_predictions_tensor(preds_tensor)[0]
+        pred_split = pred_text.split()
```

Comment:
This is only for WER right? `from nemo.collections.asr.metrics.wer import word_error_rate` and then use it - it can switch from WER to CER if user wants.

Reply:
If you mean to aggregate the whole references and hypotheses in a list and call this method once, then we need to drop the support for best_wer and worst_wer as we we need to know the differences for all the topk predictions. For the greedy wer and LM wer we can use this method.

I added the cer calculation to the scripts.

Reply:
Looks good then let's go with this.